pilot/pkg/security/trustdomain/bundle.go

```
@@ -0,0 +1,122 @@
+	return principalsIncludingAliases
+}
+
+func (t TrustDomainBundle) replaceTrustDomains(principal string) []string {
+	principalsForAliases := []string{}
+	principalsForAliases = append(principalsForAliases, replaceTrustDomainInPrincipal(t.TrustDomain, principal))
+	for _, tdAlias := range t.TrustDomainAliases {
+		principalsForAliases = append(principalsForAliases, replaceTrustDomainInPrincipal(tdAlias, principal))
+	}
+	return principalsForAliases
+}
+
+// replaceTrustDomainInPrincipal returns a new SPIFFE identity with the new trust domain.
+// The trust domain corresponds to the trust root of a system.
+// Refer to
+// [SPIFFE-ID](https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE-ID.md#21-trust-domain)
+// In Istio authorization, an identity is presented in the format:
+// <trust-domain>/ns/<some-namespace>/sa/<some-service-account>
+func replaceTrustDomainInPrincipal(trustDomain string, principal string) string {
+	identityParts := strings.Split(principal, "/")
+	// A valid SPIFFE identity in authorization has no SPIFFE:// prefix.
+	// It is presented as <trust-domain>/ns/<some-namespace>/sa/<some-service-account>
+	if len(identityParts) != 5 {
+		log.Errorf("Wrong SPIFFE format isStringInList: %s", principal)
+		return ""
+	}
+	return fmt.Sprintf("%s/%s", trustDomain, strings.Join(identityParts[1:], "/"))
+}
+
+// getTrustDomain returns the trust domain from an Istio authorization principal.
+// In Istio authorization, an identity is presented in the format:
+// <trust-domain>/ns/<some-namespace>/sa/<some-service-account>
+// "cluster.local/ns/default/sa/bookinfo-ratings-v2" will return "cluster.local"
+func getTrustDomain(principal string) string {
+	identityParts := strings.Split(principal, "/")
+	// A valid SPIFFE identity in authorization has no SPIFFE:// prefix.
+	// It is presented as <trust-domain>/ns/<some-namespace>/sa/<some-service-account>
+	if len(identityParts) != 5 {
+		log.Errorf("Wrong SPIFFE format isStringInList: %s", principal)
```

what is `isStringInList` in the log? also could you return the error instead of returning empty string.

---

find-replace gone wrong. Fixed

---

I think returning an empty string will beneficial here, in case of prefix or suffix, getTrustDomain might not return the right thing. I'll add a TODO and will see if it needs to be fixed in this PR #18011